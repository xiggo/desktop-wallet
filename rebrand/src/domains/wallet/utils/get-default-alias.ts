import { Networks } from "@payvo/sdk";
import { Contracts } from "@payvo/sdk-profiles";

interface GetDefaultAliasInput {
	profile: Contracts.IProfile;
	network: Networks.Network;
}

export const getDefaultAlias = ({ profile, network }: GetDefaultAliasInput): string => {
	const makeAlias = (count: number) => `${network.displayName()} #${count}`;

	const sameCoinWallets = profile.wallets().findByCoinWithNetwork(network.coin(), network.id());

	let counter = sameCoinWallets.length;

	if (counter === 0) {
		counter = 1;
	}

	while (profile.wallets().findByAlias(makeAlias(counter))) {
		counter++;
	}

	return makeAlias(counter);
};
