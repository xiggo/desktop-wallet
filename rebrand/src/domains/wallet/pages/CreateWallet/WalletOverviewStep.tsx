import React, { useEffect, useState } from "react";
import { useFormContext } from "react-hook-form";
import { Trans, useTranslation } from "react-i18next";

import { Alert } from "@/app/components/Alert";
import { Button } from "@/app/components/Button";
import { Clipboard } from "@/app/components/Clipboard";
import { Divider } from "@/app/components/Divider";
import { Header } from "@/app/components/Header";
import { Icon } from "@/app/components/Icon";
import { Toggle } from "@/app/components/Toggle";
import { toasts } from "@/app/services";
import { MnemonicList } from "@/domains/wallet/components/MnemonicList";
import { saveFile } from "@/utils/electron-utils";

export const WalletOverviewStep = () => {
	const { getValues, setValue, unregister, watch } = useFormContext();

	// getValues does not get the value of `defaultValues` on first render
	const [defaultMnemonic] = useState(() => watch("mnemonic"));
	const mnemonic = getValues("mnemonic") || defaultMnemonic;

	const [defaultWallet] = useState(() => watch("wallet"));
	const wallet = getValues("wallet") || defaultWallet;

	const { t } = useTranslation();

	useEffect(() => {
		unregister("verification");
	}, [unregister]);

	const handleDownload = async () => {
		const fileName = `${wallet.address()}.txt`;

		try {
			const filePath = await saveFile(mnemonic, fileName, {
				filters: { extensions: ["txt"], name: "Text Document" },
				returnBasename: true,
			});

			if (filePath) {
				toasts.success(
					<Trans
						i18nKey="COMMON.SAVE_FILE.SUCCESS"
						values={{ filePath }}
						components={{ bold: <strong /> }}
					/>,
				);
			}
		} catch (error) {
			toasts.error(t("COMMON.SAVE_FILE.ERROR", { error: error.message }));
		}
	};

	const handleToggleEncryption = (event: React.ChangeEvent<HTMLInputElement>) => {
		setValue("useEncryption", event.target.checked);
	};

	return (
		<section data-testid="CreateWallet__WalletOverviewStep">
			<Header title={t("WALLETS.PAGE_CREATE_WALLET.PASSPHRASE_STEP.TITLE")} />

			<Alert className="mt-6">{t("WALLETS.PAGE_CREATE_WALLET.PASSPHRASE_STEP.WARNING")}</Alert>

			<div className="mt-8 space-y-8">
				<MnemonicList mnemonic={mnemonic} />

				<Divider dashed />

				<div className="flex justify-between items-center">
					<div className="space-y-2">
						<span className="text-lg font-semibold text-theme-secondary-text">
							{t("WALLETS.PAGE_CREATE_WALLET.PASSPHRASE_STEP.COPY_OR_DOWNLOAD.TITLE")}
						</span>
						<p className="text-sm text-theme-secondary-500">
							{t("WALLETS.PAGE_CREATE_WALLET.PASSPHRASE_STEP.COPY_OR_DOWNLOAD.DESCRIPTION")}
						</p>
					</div>

					<Icon name="FrameKey" className="text-black dark:text-theme-secondary-600" size="xl" />
				</div>

				<div className="flex justify-end space-x-3 w-full">
					<Clipboard variant="button" data={mnemonic} data-testid="CreateWallet__copy">
						<Icon name="Copy" />
						<span>{t("COMMON.COPY")}</span>
					</Clipboard>

					<Button
						data-testid="CreateWallet__download"
						variant="secondary"
						className="flex items-center space-x-2"
						onClick={handleDownload}
					>
						<Icon name="ArrowDownBracket" />
						<span>{t("COMMON.DOWNLOAD")}</span>
					</Button>
				</div>

				<Divider dashed />

				<div className="flex flex-col space-y-2 w-full">
					<div className="flex justify-between items-center space-x-5">
						<span className="font-bold text-theme-secondary-text">
							{t("WALLETS.PAGE_CREATE_WALLET.PASSPHRASE_STEP.ENCRYPTION.TITLE")}
						</span>

						<span data-testid="CreateWallet__encryption">
							<Toggle
								data-testid="CreateWallet__encryption-toggle"
								defaultChecked={getValues("useEncryption")}
								onChange={handleToggleEncryption}
							/>
						</span>
					</div>

					<span className="text-sm text-theme-secondary-500 mr-12">
						{t("WALLETS.PAGE_CREATE_WALLET.PASSPHRASE_STEP.ENCRYPTION.DESCRIPTION")}
					</span>
				</div>
			</div>
		</section>
	);
};
