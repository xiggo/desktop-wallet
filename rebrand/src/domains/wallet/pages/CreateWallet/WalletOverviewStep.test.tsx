/* eslint-disable @typescript-eslint/require-await */
import { Contracts } from "@payvo/sdk-profiles";
import { renderHook } from "@testing-library/react-hooks";
import userEvent from "@testing-library/user-event";
import electron from "electron";
import React from "react";
import { FormProvider, useForm } from "react-hook-form";
import { Trans } from "react-i18next";

import { WalletOverviewStep } from "./WalletOverviewStep";
import { toasts } from "@/app/services";
import { env, getDefaultProfileId, MNEMONICS, render, screen, waitFor } from "@/utils/testing-library";

let profile: Contracts.IProfile;

const renderForm = () =>
	renderHook(() =>
		useForm({
			defaultValues: {
				mnemonic: MNEMONICS[0],
				wallet: {
					address: () => "address",
				},
			},
		}),
	);

describe("WalletOverviewStep", () => {
	beforeEach(() => {
		profile = env.profiles().findById(getDefaultProfileId());

		for (const wallet of profile.wallets().values()) {
			profile.wallets().forget(wallet.id());
		}
	});

	describe("Render step", () => {
		it("should render", async () => {
			const { result: form } = renderForm();

			jest.spyOn(electron.remote.dialog, "showSaveDialog").mockImplementation(() => ({
				filePath: "filePath",
			}));

			const { asFragment } = render(
				<FormProvider {...form.current}>
					<WalletOverviewStep />
				</FormProvider>,
			);

			expect(asFragment()).toMatchSnapshot();

			const writeTextMock = jest.fn();
			const clipboardOriginal = navigator.clipboard;
			// @ts-ignore
			navigator.clipboard = { writeText: writeTextMock };

			userEvent.click(screen.getByTestId("CreateWallet__copy"));

			await waitFor(() => expect(writeTextMock).toHaveBeenCalledWith(MNEMONICS[0]));

			// @ts-ignore
			navigator.clipboard = clipboardOriginal;
		});

		it("should show success toast on successful download", async () => {
			const { result: form } = renderForm();

			jest.spyOn(electron.remote.dialog, "showSaveDialog").mockImplementation(() => ({
				filePath: "filePath",
			}));

			const toastSpy = jest.spyOn(toasts, "success");

			render(
				<FormProvider {...form.current}>
					<WalletOverviewStep />
				</FormProvider>,
			);

			userEvent.click(screen.getByTestId("CreateWallet__download"));

			await waitFor(() => {
				expect(toastSpy).toHaveBeenCalledWith(
					<Trans
						components={{ bold: <strong /> }}
						i18nKey="COMMON.SAVE_FILE.SUCCESS"
						values={{ filePath: "filePath" }}
					/>,
				);
			});

			toastSpy.mockRestore();
		});

		it("should not show success toast on cancelled download", async () => {
			const { result: form } = renderForm();

			jest.spyOn(electron.remote.dialog, "showSaveDialog").mockImplementation(() => ({
				filePath: undefined,
			}));

			const toastSpy = jest.spyOn(toasts, "success");

			render(
				<FormProvider {...form.current}>
					<WalletOverviewStep />
				</FormProvider>,
			);

			userEvent.click(screen.getByTestId("CreateWallet__download"));

			expect(toastSpy).not.toHaveBeenCalled();

			toastSpy.mockRestore();
		});

		it("should show error toast on error", async () => {
			const { result: form } = renderForm();

			jest.spyOn(electron.remote.dialog, "showSaveDialog").mockImplementation(() => {
				throw new Error("Error");
			});

			const toastSpy = jest.spyOn(toasts, "error");

			render(
				<FormProvider {...form.current}>
					<WalletOverviewStep />
				</FormProvider>,
			);

			userEvent.click(screen.getByTestId("CreateWallet__download"));

			await waitFor(() => {
				expect(toastSpy).toHaveBeenCalledWith(expect.stringMatching(/Could not save/));
			});

			toastSpy.mockRestore();
		});
	});
});
