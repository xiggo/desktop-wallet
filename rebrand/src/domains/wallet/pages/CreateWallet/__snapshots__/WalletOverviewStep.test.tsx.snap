// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WalletOverviewStep Render step should render 1`] = `
<DocumentFragment>
  <section
    data-testid="CreateWallet__WalletOverviewStep"
  >
    <div
      class="flex justify-between items-end bg-theme-background"
    >
      <div
        class="space-y-4"
      >
        <h1
          class="mb-0 text-2xl"
          data-testid="header__title"
        >
          Your Passphrase
        </h1>
      </div>
    </div>
    <div
      class="flex overflow-hidden flex-col space-y-4 rounded-lg p-4 bg-theme-warning-50 dark:bg-theme-secondary-800 sm:space-y-0 sm:space-x-4 sm:flex-row sm:items-center mt-6"
    >
      <div
        class="h-11 w-11 flex flex-shrink-0 items-center justify-center rounded-lg text-white bg-theme-warning-600"
      >
        <div
          class="sc-bdvvtL qGXhc"
          height="20"
          width="20"
        >
          <svg>
            circle-exclamation-mark.svg
          </svg>
        </div>
      </div>
      <div
        class="flex flex-col w-full"
      >
        <div
          class="text-sm leading-relaxed break-words dark:text-theme-secondary-500"
        >
          You are responsible for storing and protecting this mnemonic passphrase offline. The Payvo Wallet cannot reveal this to you at a later time. If you lose this mnemonic passphrase, you will lose your funds.
        </div>
      </div>
    </div>
    <div
      class="mt-8 space-y-8"
    >
      <ul
        class="grid grid-cols-2 gap-x-3 gap-y-5 md:grid-cols-3 lg:grid-cols-4"
      >
        <li
          class="relative p-4 rounded border border-theme-secondary-400 dark:border-theme-secondary-700"
          data-testid="MnemonicList__item"
        >
          <span
            class="absolute top-0 left-0 px-1 text-xs font-semibold transform translate-x-2 -translate-y-2 bg-theme-background text-theme-secondary-700"
          >
            1
          </span>
          <span>
            skin
          </span>
        </li>
        <li
          class="relative p-4 rounded border border-theme-secondary-400 dark:border-theme-secondary-700"
          data-testid="MnemonicList__item"
        >
          <span
            class="absolute top-0 left-0 px-1 text-xs font-semibold transform translate-x-2 -translate-y-2 bg-theme-background text-theme-secondary-700"
          >
            2
          </span>
          <span>
            fortune
          </span>
        </li>
        <li
          class="relative p-4 rounded border border-theme-secondary-400 dark:border-theme-secondary-700"
          data-testid="MnemonicList__item"
        >
          <span
            class="absolute top-0 left-0 px-1 text-xs font-semibold transform translate-x-2 -translate-y-2 bg-theme-background text-theme-secondary-700"
          >
            3
          </span>
          <span>
            security
          </span>
        </li>
        <li
          class="relative p-4 rounded border border-theme-secondary-400 dark:border-theme-secondary-700"
          data-testid="MnemonicList__item"
        >
          <span
            class="absolute top-0 left-0 px-1 text-xs font-semibold transform translate-x-2 -translate-y-2 bg-theme-background text-theme-secondary-700"
          >
            4
          </span>
          <span>
            mom
          </span>
        </li>
        <li
          class="relative p-4 rounded border border-theme-secondary-400 dark:border-theme-secondary-700"
          data-testid="MnemonicList__item"
        >
          <span
            class="absolute top-0 left-0 px-1 text-xs font-semibold transform translate-x-2 -translate-y-2 bg-theme-background text-theme-secondary-700"
          >
            5
          </span>
          <span>
            coin
          </span>
        </li>
        <li
          class="relative p-4 rounded border border-theme-secondary-400 dark:border-theme-secondary-700"
          data-testid="MnemonicList__item"
        >
          <span
            class="absolute top-0 left-0 px-1 text-xs font-semibold transform translate-x-2 -translate-y-2 bg-theme-background text-theme-secondary-700"
          >
            6
          </span>
          <span>
            hurdle
          </span>
        </li>
        <li
          class="relative p-4 rounded border border-theme-secondary-400 dark:border-theme-secondary-700"
          data-testid="MnemonicList__item"
        >
          <span
            class="absolute top-0 left-0 px-1 text-xs font-semibold transform translate-x-2 -translate-y-2 bg-theme-background text-theme-secondary-700"
          >
            7
          </span>
          <span>
            click
          </span>
        </li>
        <li
          class="relative p-4 rounded border border-theme-secondary-400 dark:border-theme-secondary-700"
          data-testid="MnemonicList__item"
        >
          <span
            class="absolute top-0 left-0 px-1 text-xs font-semibold transform translate-x-2 -translate-y-2 bg-theme-background text-theme-secondary-700"
          >
            8
          </span>
          <span>
            emotion
          </span>
        </li>
        <li
          class="relative p-4 rounded border border-theme-secondary-400 dark:border-theme-secondary-700"
          data-testid="MnemonicList__item"
        >
          <span
            class="absolute top-0 left-0 px-1 text-xs font-semibold transform translate-x-2 -translate-y-2 bg-theme-background text-theme-secondary-700"
          >
            9
          </span>
          <span>
            heart
          </span>
        </li>
        <li
          class="relative p-4 rounded border border-theme-secondary-400 dark:border-theme-secondary-700"
          data-testid="MnemonicList__item"
        >
          <span
            class="absolute top-0 left-0 px-1 text-xs font-semibold transform translate-x-2 -translate-y-2 bg-theme-background text-theme-secondary-700"
          >
            10
          </span>
          <span>
            brisk
          </span>
        </li>
        <li
          class="relative p-4 rounded border border-theme-secondary-400 dark:border-theme-secondary-700"
          data-testid="MnemonicList__item"
        >
          <span
            class="absolute top-0 left-0 px-1 text-xs font-semibold transform translate-x-2 -translate-y-2 bg-theme-background text-theme-secondary-700"
          >
            11
          </span>
          <span>
            exact
          </span>
        </li>
        <li
          class="relative p-4 rounded border border-theme-secondary-400 dark:border-theme-secondary-700"
          data-testid="MnemonicList__item"
        >
          <span
            class="absolute top-0 left-0 px-1 text-xs font-semibold transform translate-x-2 -translate-y-2 bg-theme-background text-theme-secondary-700"
          >
            12
          </span>
          <span>
            rather
          </span>
        </li>
        <li
          class="relative p-4 rounded border border-theme-secondary-400 dark:border-theme-secondary-700"
          data-testid="MnemonicList__item"
        >
          <span
            class="absolute top-0 left-0 px-1 text-xs font-semibold transform translate-x-2 -translate-y-2 bg-theme-background text-theme-secondary-700"
          >
            13
          </span>
          <span>
            code
          </span>
        </li>
        <li
          class="relative p-4 rounded border border-theme-secondary-400 dark:border-theme-secondary-700"
          data-testid="MnemonicList__item"
        >
          <span
            class="absolute top-0 left-0 px-1 text-xs font-semibold transform translate-x-2 -translate-y-2 bg-theme-background text-theme-secondary-700"
          >
            14
          </span>
          <span>
            feature
          </span>
        </li>
        <li
          class="relative p-4 rounded border border-theme-secondary-400 dark:border-theme-secondary-700"
          data-testid="MnemonicList__item"
        >
          <span
            class="absolute top-0 left-0 px-1 text-xs font-semibold transform translate-x-2 -translate-y-2 bg-theme-background text-theme-secondary-700"
          >
            15
          </span>
          <span>
            era
          </span>
        </li>
        <li
          class="relative p-4 rounded border border-theme-secondary-400 dark:border-theme-secondary-700"
          data-testid="MnemonicList__item"
        >
          <span
            class="absolute top-0 left-0 px-1 text-xs font-semibold transform translate-x-2 -translate-y-2 bg-theme-background text-theme-secondary-700"
          >
            16
          </span>
          <span>
            leopard
          </span>
        </li>
        <li
          class="relative p-4 rounded border border-theme-secondary-400 dark:border-theme-secondary-700"
          data-testid="MnemonicList__item"
        >
          <span
            class="absolute top-0 left-0 px-1 text-xs font-semibold transform translate-x-2 -translate-y-2 bg-theme-background text-theme-secondary-700"
          >
            17
          </span>
          <span>
            grocery
          </span>
        </li>
        <li
          class="relative p-4 rounded border border-theme-secondary-400 dark:border-theme-secondary-700"
          data-testid="MnemonicList__item"
        >
          <span
            class="absolute top-0 left-0 px-1 text-xs font-semibold transform translate-x-2 -translate-y-2 bg-theme-background text-theme-secondary-700"
          >
            18
          </span>
          <span>
            tide
          </span>
        </li>
        <li
          class="relative p-4 rounded border border-theme-secondary-400 dark:border-theme-secondary-700"
          data-testid="MnemonicList__item"
        >
          <span
            class="absolute top-0 left-0 px-1 text-xs font-semibold transform translate-x-2 -translate-y-2 bg-theme-background text-theme-secondary-700"
          >
            19
          </span>
          <span>
            gift
          </span>
        </li>
        <li
          class="relative p-4 rounded border border-theme-secondary-400 dark:border-theme-secondary-700"
          data-testid="MnemonicList__item"
        >
          <span
            class="absolute top-0 left-0 px-1 text-xs font-semibold transform translate-x-2 -translate-y-2 bg-theme-background text-theme-secondary-700"
          >
            20
          </span>
          <span>
            power
          </span>
        </li>
        <li
          class="relative p-4 rounded border border-theme-secondary-400 dark:border-theme-secondary-700"
          data-testid="MnemonicList__item"
        >
          <span
            class="absolute top-0 left-0 px-1 text-xs font-semibold transform translate-x-2 -translate-y-2 bg-theme-background text-theme-secondary-700"
          >
            21
          </span>
          <span>
            lawsuit
          </span>
        </li>
        <li
          class="relative p-4 rounded border border-theme-secondary-400 dark:border-theme-secondary-700"
          data-testid="MnemonicList__item"
        >
          <span
            class="absolute top-0 left-0 px-1 text-xs font-semibold transform translate-x-2 -translate-y-2 bg-theme-background text-theme-secondary-700"
          >
            22
          </span>
          <span>
            sight
          </span>
        </li>
        <li
          class="relative p-4 rounded border border-theme-secondary-400 dark:border-theme-secondary-700"
          data-testid="MnemonicList__item"
        >
          <span
            class="absolute top-0 left-0 px-1 text-xs font-semibold transform translate-x-2 -translate-y-2 bg-theme-background text-theme-secondary-700"
          >
            23
          </span>
          <span>
            vehicle
          </span>
        </li>
        <li
          class="relative p-4 rounded border border-theme-secondary-400 dark:border-theme-secondary-700"
          data-testid="MnemonicList__item"
        >
          <span
            class="absolute top-0 left-0 px-1 text-xs font-semibold transform translate-x-2 -translate-y-2 bg-theme-background text-theme-secondary-700"
          >
            24
          </span>
          <span>
            coin
          </span>
        </li>
      </ul>
      <div
        class="Divider-sc-17ehhxk-0 FRyyK border-theme-secondary-300 dark:border-theme-secondary-800"
        type="horizontal"
      />
      <div
        class="flex justify-between items-center"
      >
        <div
          class="space-y-2"
        >
          <span
            class="text-lg font-semibold text-theme-secondary-text"
          >
            Copy or Download Mnemonic Passphrase
          </span>
          <p
            class="text-sm text-theme-secondary-500"
          >
            You can copy or download your mnemonic, but store it safely.
          </p>
        </div>
        <div
          class="sc-bdvvtL kCgesw text-black dark:text-theme-secondary-600"
          height="40"
          width="40"
        >
          <svg>
            frame-key.svg
          </svg>
        </div>
      </div>
      <div
        class="flex justify-end space-x-3 w-full"
      >
        <div
          class="relative"
        >
          <button
            class="ClipboardButton__StyledButton-sc-cammp1-0 bAhMDA"
            data-testid="CreateWallet__copy"
            type="button"
          >
            <div
              class="flex items-center space-x-2"
            >
              <div
                class="sc-bdvvtL fCrhkg"
                height="16"
                width="16"
              >
                <svg>
                  copy.svg
                </svg>
              </div>
              <span>
                Copy
              </span>
            </div>
          </button>
        </div>
        <button
          class="OriginalButton__StyledButton-sc-o0fs2j-0 cxgwmk flex items-center space-x-2"
          data-testid="CreateWallet__download"
          type="button"
        >
          <div
            class="flex items-center space-x-2"
          >
            <div
              class="sc-bdvvtL fCrhkg"
              height="16"
              width="16"
            >
              <svg>
                arrow-down-bracket.svg
              </svg>
            </div>
            <span>
              Download
            </span>
          </div>
        </button>
      </div>
      <div
        class="Divider-sc-17ehhxk-0 FRyyK border-theme-secondary-300 dark:border-theme-secondary-800"
        type="horizontal"
      />
      <div
        class="flex flex-col space-y-2 w-full"
      >
        <div
          class="flex justify-between items-center space-x-5"
        >
          <span
            class="font-bold text-theme-secondary-text"
          >
            Use Wallet Encryption
          </span>
          <span
            data-testid="CreateWallet__encryption"
          >
            <label
              class="Toggle__Wrapper-sc-6s3wfu-0 efNWJo"
            >
              <input
                class="Toggle__Input-sc-6s3wfu-1 WwplI"
                data-testid="CreateWallet__encryption-toggle"
                type="checkbox"
              />
              <div
                class="Toggle__Handle-sc-6s3wfu-2 cobHFC"
              >
                <span
                  class="Toggle__HandleInner-sc-6s3wfu-3 ddepvC"
                />
              </div>
            </label>
          </span>
        </div>
        <span
          class="text-sm text-theme-secondary-500 mr-12"
        >
          This will allow you to set an encryption password during creation to use instead of your mnemonic.
        </span>
      </div>
    </div>
  </section>
</DocumentFragment>
`;
